lib/quicrq.c

```diff
@@ -1828,22 +1931,35 @@ int quicrq_callback(picoquic_cnx_t* cnx,
         case picoquic_callback_stream_data:
         case picoquic_callback_stream_fin:
             /* Data arrival on stream #x, maybe with fin mark */
-            if (stream_ctx == NULL) {
-                /* Retrieve, or create and initialize stream context */
+            if ((stream_id & 2) == 0 && stream_ctx == NULL) {
+                /* Retrieve, or create and initialize stream context for control channel */
                 stream_ctx = quicrq_find_or_create_stream(stream_id, cnx_ctx, 1);
-            }
+                if (stream_ctx == NULL) {
+                    /* Internal error */
+                    (void)picoquic_reset_stream(cnx, stream_id, QUICRQ_ERROR_INTERNAL);
+                    return(-1);
+                }
 
-            if (stream_ctx == NULL) {
-                /* Internal error */
-                (void)picoquic_reset_stream(cnx, stream_id, QUICRQ_ERROR_INTERNAL);
-                return(-1);
-            }
-            else {
-                ret = quicrq_receive_stream_data(stream_ctx, bytes, length, (fin_or_event == picoquic_callback_stream_fin));
+                /* we consider unidir streams as warp/rush streams. may be we need something more explicit */
+                ret = quicrq_receive_stream_data(stream_ctx, bytes, length,(fin_or_event == picoquic_callback_stream_fin));
+
+            } else {
+
+                uni_stream_ctx = quicrq_find_or_create_uni_stream(stream_id, cnx_ctx, 1);
+                if (uni_stream_ctx == NULL) {
+                    /* Internal error */
+                    (void) picoquic_reset_stream(cnx, stream_id, QUICRQ_ERROR_INTERNAL);
+                    return (-1);
```

Comment:
The receiver cannot reset a uni stream, only the sender can. The receiver could try to send "stop sending". Or you could treat that as a fatal error and close the connection.